File: src/graph_adj.py
class GraphListAdj:
    def __init__(self):
        self.adj = []

    def add_vertex(self):
        self.adj.append([])

    def add_edge(self, u, v):
        max_id = max(u, v)
        while len(self.adj) <= max_id:
            self.add_vertex()
        self.adj[u].append(v)
        self.adj[v].append(u)

    def visit_edge(self, u, v):
        if u >= len(self.adj) or v >= len(self.adj):
            return 0
        return 1 if v in self.adj[u] else 0

    def degree(self, v):
        if v >= len(self.adj):
            return 0
        return len(self.adj[v])

    def remove_edge(self, u, v):
        if self.visit_edge(u, v):
            self.adj[u].remove(v)
            self.adj[v].remove(u)

File: src/graph_adj.cpp
#include <iostream>
#include <vector>

using namespace std;

class Graph
{
private:
    vector<vector<int>> adj;

public:
    Graph() {}

    void addVertex()
    {
        adj.push_back(vector<int>());
    }

    void addEdge(int u, int v)
    {
        int maxId = max(u, v);
        while (adj.size() <= maxId)
        {
            addVertex();
        }
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    void deleteEdge(int u, int v)
    {
        if (u >= adj.size() || v >= adj.size())
            return;
        auto &vecU = adj[u];
        auto &vecV = adj[v];
        vecU.erase(remove(vecU.begin(), vecU.end(), v), vecU.end());
        vecV.erase(remove(vecV.begin(), vecV.end(), u), vecV.end());
    }

    int visitEdge(int u, int v)
    {
        if (u >= adj.size() || v >= adj.size())
            return 0;
        for (auto x : adj[u])
        {
            if (x == v)
                return x;
        }
        return 0;
    }

    int degree(int v)
    {
        if (v >= adj.size())
            return 0;
        return adj[v].size();
    }
};
